<a href="https://colab.research.google.com/github/Gattalraouf/Feature-Selection-Wrapper-methods/blob/master/Feature_Selection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing Data & libraries

In [0]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

from google.colab import files
uploaded = files.upload()

Saving house-votes-84.data to house-votes-84.data


In [0]:
header = ["Class Name", "handicapped-infants", "water-project-cost-sharing","adoption-of-the-budget-resolution",
          "physician-fee-freeze","el-salvador-aid","religious-groups-in-schools","anti-satellite-test-ban","aid-to-nicaraguan-contras",
          "mx-missile","immigration","synfuels-corporation-cutback","education-spending","superfund-right-to-sue","crime","duty-free-exports",
          "export-administration-act-south-africa"]
import io
data = pd.read_csv(io.BytesIO(uploaded['house-votes-84.data']), names=header)
#notes = pd.read_csv('datasets/marks.txt', names=header)

# Exploring Data :
Size : 435 line of data

Number of features : 17 feature

Number of Null elements : 

* Class Name :                                 0
* handicapped-infants :                       12
* water-project-cost-sharing :                48
* adoption-of-the-budget-resolution :         11
* physician-fee-freeze :                      11
* el-salvador-aid :                           15
* religious-groups-in-schools :               11
* anti-satellite-test-ban :                   14
* aid-to-nicaraguan-contras :                  15
* mx-missile :                                22
* immigration :                                7
* synfuels-corporation-cutback :              21
* education-spending :                        31
* superfund-right-to-sue :                    25
* crime :                                     17
* duty-free-exports :                         28
* export-administration-act-south-africa :   104

In [0]:
# Exploring Data
np.shape(data)

(435, 17)

In [0]:
data.head(5)

Class Name  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                      ?
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [0]:
# Changing indexes
variable_changes ={'y':1,'n':0,'?':None}
data =  data.replace(variable_changes)

In [0]:
data.isnull().sum()

Class Name                                  0
handicapped-infants                        12
water-project-cost-sharing                 48
adoption-of-the-budget-resolution          11
physician-fee-freeze                       11
el-salvador-aid                            15
religious-groups-in-schools                11
anti-satellite-test-ban                    14
aid-to-nicaraguan-contras                  15
mx-missile                                 22
immigration                                 7
synfuels-corporation-cutback               21
education-spending                         31
superfund-right-to-sue                     25
crime                                      17
duty-free-exports                          28
export-administration-act-south-africa    104
dtype: int64

In [0]:
data.head(5)

Class Name  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [0]:
# droping the null value lines
new_data = data.dropna(axis = 0, how ='any')

# Exploring Data After Removing the null values
Size : 232

Number of features : 17

Null values : All 0

In [0]:
new_data.isnull().sum()

Class Name                                0
handicapped-infants                       0
water-project-cost-sharing                0
adoption-of-the-budget-resolution         0
physician-fee-freeze                      0
el-salvador-aid                           0
religious-groups-in-schools               0
anti-satellite-test-ban                   0
aid-to-nicaraguan-contras                 0
mx-missile                                0
immigration                               0
synfuels-corporation-cutback              0
education-spending                        0
superfund-right-to-sue                    0
crime                                     0
duty-free-exports                         0
export-administration-act-south-africa    0
dtype: int64

In [0]:
new_data.head(5)

Class Name  ...  export-administration-act-south-africa
5     democrat  ...                                     1.0
8   republican  ...                                     1.0
19    democrat  ...                                     1.0
23    democrat  ...                                     1.0
25    democrat  ...                                     1.0

[5 rows x 17 columns]

In [0]:
np.shape(new_data)

(232, 17)

# Logistic Regression

Accuaracy : 0.916666...

In [0]:
# replacing "democrat" & "republican"
variable_map ={'republican':0,'democrat':1}
final_data = new_data.replace(variable_map)

In [0]:
# Diviser X,y
X = final_data.values[:, 1:16]
y = final_data.values[:, 0]

In [0]:
# Applying logistic regression
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score 
from sklearn.model_selection import train_test_split

## Décider de la taille des données pour chaque set
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.1)

print(X_train.shape)
print(X_test.shape)

# Créer le modèle 
model = LogisticRegression(solver="lbfgs")

# Entraîner le modèle 
model.fit(X_train, y_train)

# Prédire les classes 
predicted_classes = model.predict(X_test)

# Calculer le score du modèle 
accuracy = accuracy_score(y_test.flatten(),predicted_classes)

print(accuracy)

(208, 15)
(24, 15)
0.9166666666666666


# Wrapper feature Selection (RFE)
Recursive feature elimination performs a greedy search to find the best performing feature subset. It iteratively creates models and determines the best or the worst performing feature at each iteration. It constructs the subsequent models with the left features until all the features are explored. It then ranks the features based on the order of their elimination. In the worst case, if a dataset contains N number of features RFE will do a greedy search for 2N combinations of features.

After Testing with the dataset results : 

Accuaracy : 0.958333.....4

Number of chosen features : 1

Taken Time : 0.06500816345214844

In [0]:

# Prédire les classes 
h = fit.predict(X_test)

# Calculer le score du modèle 
accuracy = accuracy_score(y_test,h)

print(accuracy)

0.9583333333333334


In [0]:
# Wrapper feature Selection 

# Import your necessary dependencies
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression

import time
start = time.time()

# Feature extraction
model_rfe = LogisticRegression(solver="lbfgs")
rfe = RFE(model_rfe, 
         1)


fit = rfe.fit(X_train, y_train)
end = time.time()
print("Taken Time :")
print(end - start)

print("Num Features: %s" % (fit.n_features_))
print("Selected Features: %s" % (fit.support_))
print("Feature Ranking: %s" % (fit.ranking_))

Taken Time :
0.06500816345214844
Num Features: 1
Selected Features: [False False False  True False False False False False False False False
 False False False]
Feature Ranking: [14 10  2  1  4 13 11 12  8  6  3  5 15  7  9]


# Wrapper feature Selection (RFE after removing the main feature)

Accuracy : 0.875

Number of chosen  features : 4

Taken Time : 0.05390310287475586

In [0]:
# Diviser X,y en ignorant 
rfe_data = final_data.drop(['physician-fee-freeze'], axis=1)

In [0]:
# Diviser X,y
X_rfe = rfe_data.values[:, 1:15]
y_rfe = rfe_data.values[:, 0]

## Décider de la taille des données pour chaque set
X_rfe_train, X_rfe_test, y_rfe_train, y_rfe_test = train_test_split(X_rfe,y_rfe,test_size=0.1)

In [0]:
# Feature extraction
start = time.time()

model = LogisticRegression(solver="lbfgs")
rfe = RFE(model, 
         4)
fit = rfe.fit(X_rfe_train, y_rfe_train)

end = time.time()
print("Taken Time :")
print(end - start)

print("Num Features: %s" % (fit.n_features_))
print("Selected Features: %s" % (fit.support_))
print("Feature Ranking: %s" % (fit.ranking_))


Taken Time :
0.05390310287475586
Num Features: 4
Selected Features: [False False  True  True False False False False False  True False False
  True False]
Feature Ranking: [10 11  1  1  6  7  9  5  3  1  2  8  1  4]


In [0]:
# Prédire les classes 
h = fit.predict(X_rfe_test)

# Calculer le score du modèle 
accuracy = accuracy_score(y_rfe_test,h)

print(accuracy)

0.875


# Wrapper feature Selection (Step Forward Feature Selection)

The procedure starts with an empty set of features [reduced set]. The best of the original features is determined and added to the reduced set. At each subsequent iteration, the best of the remaining original attributes is added to the set. 

After Testing with the dataset results : 

Accuracy : 0.9583333...4

Number of chosen features : 1

Taken Time : 0.19947099685668945

In [0]:
from mlxtend.feature_selection import SequentialFeatureSelector as sfs

start = time.time()
# Créer le modèle 
model = LogisticRegression(solver="lbfgs")

# Build step forward feature selection
sfs1 = sfs(model,
           k_features=1,
           forward=True,
           floating=False,
           verbose=2,
           scoring='accuracy',
           cv=5)

# Perform SFFS
sfs1 = sfs1.fit(X_train, y_train)

end = time.time()
print("Taken Time :")
print(end - start)

# Which features?
feat_cols = list(sfs1.k_feature_idx_)
print(feat_cols)

Taken Time :
0.19947099685668945
[3]


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  15 out of  15 | elapsed:    0.2s finished

[2019-07-01 10:05:02] Features: 1/1 -- score: 0.9714285714285715

In [0]:
# Generate the new subsets based on the selected features
X_train_sfs = sfs1.transform(X_train)
X_test_sfs = sfs1.transform(X_test)

# Fit the estimator using the new feature subset
# and make a prediction on the test data
model.fit(X_train_sfs, y_train)
h = model.predict(X_test_sfs)

# Calculer le score du modèle 
accuracy = accuracy_score(y_test,h)

print(accuracy)

0.9583333333333334


# Wrapper feature Selection (Step Forward Feature Selection after removing the main feature)

Accuracy : 0.791666...

Number of chosen  features : 8

Taken Time : 1.4411256313323975

In [0]:
# Créer le modèle 

start = time.time()

model1 = LogisticRegression(solver="lbfgs")

# Build step forward feature selection
sfs1 = sfs(model,
           k_features=8,
           forward=True,
           floating=False,
           verbose=2,
           scoring='accuracy',
           cv=5)

# Perform SFFS
sfs1 = sfs1.fit(X_rfe_train, y_rfe_train)

end = time.time()
print("Taken Time :")
print(end - start)

# Which features?
feat_cols = list(sfs1.k_feature_idx_)
print(feat_cols)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  14 out of  14 | elapsed:    0.2s finished

[2019-07-01 10:05:41] Features: 1/8 -- score: 0.8842857142857143[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  13 out of  13 | elapsed:    0.2s finished

[2019-07-01 10:05:41] Features: 2/8 -- score: 0.8988095238095237[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  12 out of  12 | elapsed:    0.2s finished

[2019-07-01 10:05:41] Features: 3/8 -- score: 0.8988095238095237[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 

Taken Time :
1.4411256313323975
[0, 1, 3, 4, 5, 7, 10, 12]


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    0.2s finished

[2019-07-01 10:05:42] Features: 8/8 -- score: 0.9038095238095238

In [0]:
# Generate the new subsets based on the selected features
X_train_sfs = sfs1.transform(X_rfe_train)
X_test_sfs = sfs1.transform(X_rfe_test)

# Fit the estimator using the new feature subset
# and make a prediction on the test data
model.fit(X_train_sfs, y_rfe_train)
h = model.predict(X_test_sfs)

# Calculer le score du modèle 
accuracy = accuracy_score(y_rfe_test,h)

print(accuracy)

0.7916666666666666


# Logistic Regression (After removing the main feature)

Accuaracy : 0.875

In [0]:

# Créer le modèle 
model = LogisticRegression(solver="lbfgs")

# Entraîner le modèle 
model.fit(X_rfe_train, y_rfe_train)

# Prédire les classes 
predicted_classes = model.predict(X_rfe_test)

# Calculer le score du modèle 
accuracy = accuracy_score(y_rfe_test.flatten(),predicted_classes)

print(accuracy)

0.875
